include/dslash_helper.cuh

```diff
@@ -429,6 +491,129 @@ namespace quda
+      int dim;
+      int threadl;
+      int threads_my_dir;
+      switch (myblockidx / blocks_per_dim) {
+      case 0: dim = arg.dim_map[0]; break;
+      case 1: dim = arg.dim_map[1]; break;
+      case 2: dim = arg.dim_map[2]; break;
+      case 3: dim = arg.dim_map[3]; break;
+      default: dim = -1;
+      }
+
+      switch (dim) {
+      case 0:
+        threads_my_dir = (arg.threadDimMapUpper[0] - arg.threadDimMapLower[0]) / 2;
+        threadl = arg.threadDimMapLower[0];
+        break;
+      case 1:
+        threads_my_dir = (arg.threadDimMapUpper[1] - arg.threadDimMapLower[1]) / 2;
+        threadl = arg.threadDimMapLower[1];
+        break;
+      case 2:
+        threads_my_dir = (arg.threadDimMapUpper[2] - arg.threadDimMapLower[2]) / 2;
+        threadl = arg.threadDimMapLower[2];
+        break;
+      case 3:
+        threads_my_dir = (arg.threadDimMapUpper[3] - arg.threadDimMapLower[3]) / 2;
+        threadl = arg.threadDimMapLower[3];
+        break;
+      default: threadl = 0; threads_my_dir = 0;
+      }
+      int dimdir = 2 * dim + dir;
+
+      constexpr bool shmembarrier = true; // always true for now (arg.shmem & 16);
+
+      if (shmembarrier) {
+
+        if (shmem_interiordone && threadIdx.x == blockDim.x - 1 && threadIdx.y == 0 && threadIdx.z == 0) {
+          long tst_val = interior_done.load(cuda::std::memory_order_relaxed);
+          while (tst_val < arg.counter - 1) {
+            interior_done.compare_exchange_strong(tst_val, arg.counter - 1, cuda::std::memory_order_relaxed,
+                                                  cuda::std::memory_order_relaxed);
+          }
+          interior_done.wait(arg.counter - 1, cuda::std::memory_order_acquire);
+        }
+        if (threadIdx.x < 8 && threadIdx.y == 0 && threadIdx.z == 0) {
+          // spin for my direction
+          bool spin = threadIdx.x == dimdir ? true : false;
+          // figure out which other directions also to spin for (to make corners work)
+          switch (dim) {
+          case 3:
+            if (arg.commDim[3]) {
+              spin = threadIdx.x / 2 < 3 ? arg.commDim[2] : spin;
+              spin = threadIdx.x / 2 < 2 ? arg.commDim[1] : spin;
+              spin = threadIdx.x / 2 < 1 ? arg.commDim[0] : spin;
+            } else {
+              spin = false;
+            }
+            break;
+          case 2:
+            if (arg.commDim[2]) {
+              if (arg.commDim[1]) spin = threadIdx.x / 2 < 2 ? true : spin;
+              if (arg.commDim[0]) spin = threadIdx.x / 2 < 1 ? true : spin;
+            }
+            break;
+          case 1:
+            if (arg.commDim[1]) {
+              if (arg.commDim[0]) spin = threadIdx.x / 2 < 1 ? true : spin;
+            }
+            break;
+          case 0: break;
+          }
+
+          if (getNeighborRank(threadIdx.x, arg) >= 0) {
+            if (spin) {
+              while (arg.counter > *(arg.sync_arr + threadIdx.x)) { __nanosleep(10); };
```

Comment:
Curious, how did you arrive at a value of 10 here?

Reply:
No strong motivation for that here. Worth looking tuning that a bit.